lib/index.d.ts

```diff
@@ -0,0 +1,185 @@
+type i53 = number;
+type i24 = number;
+
+type Ty = Array<i53>;
+
+type Column = Array<unknown>;
+
+type Archetype = {
+	id: number,
+	edges: {
+		[key: i53]: {
+			add: Archetype,
+			remove: Archetype,
+		},
+	},
+	types: Ty,
+	type: string | number,
+	entities: Array<number>,
+	columns: Array<Column>,
+	records: { [key: number]: number },
+}
+
+type ArchetypeMap = {
+	cache: Array<ArchetypeRecord>,
+	first: ArchetypeMap,
+	second: ArchetypeMap,
+	parent: ArchetypeMap,
+	size: number,
+}
+
+type ArchetypeRecord = number;
+
+type EntityIndex = {
+    dense: {
+        [key: i24]: i53
+    };
+    sparse: {
+        [key: i53]: Record
+    }
+}
+
+type Record = {
+	archetype: Archetype,
+	row: number,
+	dense: i24,
+	componentRecord: ArchetypeMap,
+}
+
+type Query<T extends unknown[]> = {
+    without: (...components: Entity[]) => Query<T>;
+} & IterableFunction<LuaTuple<[Entity, ...T]>>
+
+// Utility Types
+export type Entity<T = unknown> = number & { __nominal_type_dont_use: T }
+export type EntityType<T> = T extends Entity<infer A> ? A : never;
+export type InferComponents<A extends Entity[]> = {
+    [K in keyof A]: EntityType<A[K]>
+};
+type PossiblyUndefinedPack<T extends unknown[]> = {

```

Comment:
The name is quite verbose, It should be renamed NullablePack or Nullable for short. Or maybe even Nilable.

Reply:
Renamed to Nullable. In the typescript ecosystem it doesn't make sense to call it Nilable.